### PR TITLE
Cleanup generated ACLs when their associated workflow is removed

### PR DESCRIPTION
**Pull request by hsong-rh**
Add `WorkflowDeleteService` to cleanup ACLs when workflow is destroyed.

### CODE REVIEW

File: app/services/workflow_delete_service.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+class WorkflowDeleteService
+  attr_accessor :workflow
+
+  def initialize(workflow_id)
+    self.workflow = Workflow.find(workflow_id)
+  end
+
+  def cleanup
```
Comment (bzwei) [inline app/services/workflow_delete_service.rb]: just called `delete`

Comment (bzwei) [reply app/services/workflow_delete_service.rb]: also need to delete the entries in tag links table, but we also need to delete the actual tag remotely.

File: spec/services/workflow_delete_service_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+  subject { described_class.new(workflow.id) }
+
+  context 'when update' do
+    before do
+      allow(AccessProcessService).to receive(:new).and_return(aps)
+      allow(aps).to receive(:remove_resource_from_groups)
+    end
+
+    it 'returns status code 204' do
```
Comment (bzwei) [inline spec/services/workflow_delete_service_spec.rb]: the description is wrong

File: spec/services/workflow_delete_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      Insights::API::Common::Request.with_request(RequestSpecHelper.default_request_hash) do
+        subject.cleanup
+
+        expect(Workflow.find_by(:id => workflow.id)).to be_nil
```
Comment (bzwei) [inline spec/services/workflow_delete_service_spec.rb]: also need to assert the acls being removed, and tag links removed.

File: app/services/workflow_delete_service.rb (hunk 1 above)
Comment (hsong-rh) [reply app/services/workflow_delete_service.rb]: Name is changed into `delete`. The local tag links will be cleaned up automatically. Remote tags need to implement `DeleteRemoteTags` service first. Marked `TODO` here, will add it later.

File: spec/services/workflow_delete_service_spec.rb (hunk 3 above)
Comment (hsong-rh) [reply spec/services/workflow_delete_service_spec.rb]: `acls` is stub mockup. Tests for tag links will add later.

File: spec/services/workflow_delete_service_spec.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  subject { described_class.new(workflow.id) }
+
+  context 'when delete' do
+    before do
+      allow(AccessProcessService).to receive(:new).and_return(aps)
+      allow(aps).to receive(:remove_resource_from_groups)
+    end
+
+    it 'returns nil after deletion' do
```
Comment (bzwei) [inline spec/services/workflow_delete_service_spec.rb]: `it 'deletes the workflow'`

File: app/controllers/api/v1/workflows_controller.rb [hunk 5]
```diff
@@ -34,14 +34,11 @@ def index
...
-        workflow = Workflow.find(params.require(:id))
-        workflow.destroy!
+        WorkflowDeleteService.new(params.require(:id)).delete
         head :no_content
       rescue ActiveRecord::InvalidForeignKey => e
         json_response({ :message => e.message }, :forbidden)
-      rescue ActiveRecord::RecordNotDestroyed
-        raise unless workflow.errors[:base].include?(Workflow::MSG_PROTECTED_RECORD)
-
+      rescue Exceptions::NotAuthorizedError
```
Comment (bzwei) [inline app/controllers/api/v1/workflows_controller.rb]: why to remove `RecordNotDestoryed` and why specially for `NotAuthorizedError`? Shouldn't it be commonly handled for all controller methods?

File: app/controllers/api/v1/workflows_controller.rb [hunk 6]
```diff
@@ -34,14 +34,11 @@ def index
       end
 
       def destroy
-        workflow = Workflow.find(params.require(:id))
-        workflow.destroy!
+        WorkflowDeleteService.new(params.require(:id)).delete
         head :no_content
       rescue ActiveRecord::InvalidForeignKey => e
         json_response({ :message => e.message }, :forbidden)
```
Comment (bzwei) [inline app/controllers/api/v1/workflows_controller.rb]: @lgalis suggests to improve the error message when the workflow still has requests link to it.
Maybe we should explicitly detect the association and not rely on the rails build-in protection. The logic should be placed in `WorkflowDeleteService`.

**Comment by bzwei** [comment]
close in favor of #243

**Pull request closed without merging**